### DISCUSSION TITLE
10-day playlist rotation suggestions?

### DISCUSSION THREAD

**Discussion by gregnpoole** [AzuraCast/AzuraCast#7326] [Q&A]
I'm using AzuraCast for daily group educational sessions.

We have twice daily sessions, one in the AM around 8am and another in the afternoon about 4pm.  We play the same audio for both the AM and PM sessions on that day.

There are 10 audios in our program that are played in sequence over the course of 10 days.  After the last session, we start all over again on another 10-day sequence.  We're trying to get this rotation to be on auto-pilot and not have to manually intervene to restart and/or reschedule it for the next rotation.

I have tried as many ways as I can think of to get the existing AzuraCast playlist options to do this 10-day rotation but am just not getting it accomplished.  If we had say just 7 audios, that would work perfectly as we could just schedule them play the audios Mon-Sun and the schedule would repeat without any issues.  Come to think of it, if the number of daily audios you would have is less than or greater than 7, the problem is basically the same.  How to get the rotation on auto-pilot.

Currently I'm utilizing a playlist that loads its file(s) from a folder.  Since when configured that way, the playlist will be "watching" the folder, if I replace the audio file in that folder, then it will play that file when it's next scheduled to play.  I figure that with this setup, I then just need to copy/move the appropriate daily audio into the folder at say midnight, and then the following day's session will play in the sequence.

The AzuraCast installation I'm using was installed on a Linux server (Linode/Akamai) and its running in Docker, so I think I have limited options on how to copy/move audio files around.  I remember reading somewhere on one of the AzuraCast support forums Buster saying that the directories listed for where media lives are not actually the same locations the files are actually residing in the container and to NOT go moving them around.  That sort of kills the idea I had above for moving the day's file into the playlist "watched" folder.

One of the options I have considered is to use Liquidsoap scripting to do some jiggery-pokerie behind the scenes to accomplish getting files into the watched folder.  After reading about Buster's admonitions mentioned above, I'm not sure that's the best way to do it.  Before going any further down that rabbit hole, I figured I'd ask for some educated suggestions first.

Another option I have considered is to write a shell script or small Python program to do the "figuring" of which audio should play for a given day and have it called via a cron trigger.  I think in order to go this route, I might need to re-install AzuraCast using the straightforward Linux route outside of Docker.  I'm more comfortable messing about in Linux without having to worry about a Docker container... but maybe I'm just old fashioned too!

Sometimes I can be oblivious to even the most obvious solutions, so if my description of what I'm trying to do makes sense and I'm missing the blatantly obvious solution, I'm open to your pointers.

cheers...
Greg

**Comment by gAlleb** [comment]
> so I think I have limited options on how to copy/move audio files around.

Hi, you can create or edit `docker-compose.override.yml` -> `nano /var/azuracast/docker-compose.override.yml` and mount a folder into azuracast:

```ruby
services:
  web:
    volumes:
      - /path/to/audio:/var/azuracast/stations/***YOUR-STATION***/media/your_audio
```

This way you can freely move/copy files in `/path/to/audio` folder from your home system cause it will be mounted into Azuracast.

<hr>

Indeed, it's an interesting question.

Here is my quick idea, however it could be done much elegantly - just need time to think about it.

1. We create 2 identical playlists with 10 audio files and call them `10_8am` and `10_16pm`.
2. We make them Advanced playlists.
3. Go to Broadcasting -> Edit Liquidsoap Configuration -> Third Box and paste:
```
playlist_10_8am = playlist(mode="normal", loop=true, "/var/azuracast/stations/radio/playlists/playlist_10_8am.m3u")

playlist_10_16pm = playlist(mode="normal", loop=true, "/var/azuracast/stations/radio/playlists/playlist_10_16pm.m3u")

radio = mksafe(switch(track_sensitive=false,
      [(predicate.once({ 8h00m }), playlist_10_8am), 
       (predicate.once({ 16h00m }), playlist_10_16pm) 
       ]
    ))
```

So now we have 2 playlists, each one contains 10 identical files, they are sequential (mode='normal'), they are not watched, so we can't make changes to them on-the-fly..
We have a `switch`, which plays everyday 1 file from 1st playlist at 8:00 and 1 file from 2nd playlist at 16:00 and silence in between. When 10 days are gone - it all starts over).

I see how it's not ideal at all) Cause we have to start Azuracast specifically on a day when first 10-day program starts. Some thinking should be made)

<hr>

Giving it a second thought - It could be achieved with inner Azuracast scheduling (however I'm not sure how robust it would be, but its possible that it would work):

1. We create 2 identical playlists with 10 audio files and call them `10_8am` and `10_16pm`.
2. Each playlist should be "Sequential"
<img width="500" alt="image" src="https://github.com/user-attachments/assets/e21d1a66-61ed-4453-abb1-a315866fdc88">

3. We schedule them (E.g. `10_8am` playlist, make an interval a bit larger then actual track):
<img width="500" alt="image" src="https://github.com/user-attachments/assets/05f8da6d-f83f-4427-82e3-548e5355d9d9">

4. We check `interupt` option and `play only one audio at scheduled time` option.
<img width="626" alt="image" src="https://github.com/user-attachments/assets/b3724fda-4f00-44d0-ae16-5eb98b6fcc56">

This should mimic the first option I've proposed and make available reordering playlists on the fly. To make it robust we can switch to `Manual AutoDJ Mode` in Profile settings.

**Reply by gregnpoole** [reply]
OK I can see that I was a little unclear about the playlists I described in my original post! Thought I read & re-read the description before pressing send...

The point I think I missed is that each of the 10 audios we're using are each different to each other. So 10 individual, unique audios, or we might say lessons. On day one, play audio #1 twice that day, once in the morning and once in the afternoon. On day two, play audio #2 twice. Day three, audio #3, etc. So it takes 10 days to work through the sequence of unique audios. At the end of playing the 10 audios over 10 days, we rotate to the beginning and start over.

Solution #1 - override docker to mount a local folder into AzuraCast

Thanks for pointing me to this. I vaguely recall seeing this technique elsewhere but wasn't thinking about my problem at the time. This might work for us. I would like to keep everything on the remote Linode server instead of having a volume mounted on my local hard disk, but still this is something worth having in the pocket.

Solution #2 - advanced playlists and Liquidsoap programming

Also a nice creative solution, using advanced playlists. These suggestions are getting the creative juices flowing...

I'm new to Liquidsoap, having just taken a deep dive into it over the weekend. Just starting to feel comfortable with it and need much more reading. Was trying to inject some LS code using the page you have suggested here but was having a hard time seeing if anything I did was working or not.  I wasn't using an advanced playlist, so that's why things weren't working.

After some investigation, I could see that just having AzuraCast load the code by reloading the configuration was the same as compiling the code. If the reload worked OK, the code compiled OK, if not, errors were shown. I was using log() statements to debug the code and loading the resulting Liquidsoap log file into an editor and searching for my log comments. Yeah, this is debugging at it's most primitive!

As you say, this requires some manual intervention since the config file is only reloaded when the station starts up or config is reloaded.

Solution #3 - all done with some creative AzuraCast playlist manipulation

I'm still thinking about this one. At work now, so can't try anything because our network is too locked down and I can't get outside the building.

Thanks for some creative suggestions!

**Reply by gAlleb** [reply]
> OK I can see that I was a little unclear about the playlists I described in my original post! Thought I read & re-read the description before pressing send...
The point I think I missed is that each of the 10 audios we're using are each different to each other. So 10 individual, unique audios, or we might say lessons. On day one, play audio https://github.com/AzuraCast/AzuraCast/issues/1 twice that day, once in the morning and once in the afternoon. On day two, play audio https://github.com/AzuraCast/AzuraCast/issues/2 twice. Day three, audio https://github.com/AzuraCast/AzuraCast/issues/3, etc. So it takes 10 days to work through the sequence of unique audios. At the end of playing the 10 audios over 10 days, we rotate to the beginning and start over.

It's been quite clear. That is why we have `sequential mode` (mode="normal") - to go through every individual file each day.

> Thanks for pointing me to this. I vaguely recall seeing this technique elsewhere but wasn't thinking about my problem at the time. This might work for us. I would like to keep everything on the remote Linode server instead of having a volume mounted on my local hard disk, but still this is something worth having in the pocket.

This method mounts your "local" folder on remote server into container. It means that if you have a folder on your server - you can mount it inside Docker container. https://www.azuracast.com/docs/administration/docker/#mounting-a-directory-into-a-station

**Reply by gregnpoole** [reply]
OK, this makes it clearer then.  Great suggestions!

Thanks to all for all the generous help.  You all are very helpful.  AzuraCast is an amazing piece of software.
